staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
 	StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Discovery API Types
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DiscoveryAPIGroupList struct {
+	TypeMeta `json:",inline"`
+	// groups is a list of APIGroup.
+	// +listType=map
+	// +listMapKey=name
+	Groups []DiscoveryAPIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type DiscoveryAPIGroup struct {
+	TypeMeta `json:",inline"`
+	// name is the name of the group.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// versions are the versions supported in this group.
+	// This will be sorted in descending order based on the preferred version
+	// +listType=map
+	// +listMapKey=version
+	Versions []DiscoveryGroupVersion `json:"versions" protobuf:"bytes,2,rep,name=versions"`
+}
+
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensibility.
+type DiscoveryGroupVersion struct {
+	// version specifies the version in the form of "version". This is to save
+	// the clients the trouble of splitting the GroupVersion.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// resources contains the name of the resources and if they are namespaced.
+	// +listType=map
+	// +listMapKey=name
+	APIResources []DiscoveryAPIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`
+
+	// LastContacted is the last time that the apiserver has successfully reached the
+	// corresponding group version's discovery document. This will be nil if the group-version
+	// has not been aggregated yet (APIResources will be empty). To maintain consistency across scenarios with multiple
+	// apiservers, this time will be quantized down to the nearest fifteen minutes.
+	// LastContacted *time.Time `json:"lastContacted" protobuf:"bytes,3,opt,name=lastContacted"`
```

Comment:
what's the plan with this field?

Reply:
We usually avoid heartbeats because they are expensive to maintain.  While this is (currently) a poll-only API and 15 minutes is fairly long, the benefit of having this field versus having the apiservers remove stale entries or warn when something transitions to stale (status.conditions?) seems more advantageous.

Reply:
Since users are reading a cache, they need some way to evaluate freshness. We also need to be reasonably sure the caches will converge in the face of apiserver restarts etc.

Reply:
> Since users are reading a cache, they need some way to evaluate freshness. We also need to be reasonably sure the caches will converge in the face of apiserver restarts etc.

Won't this expire the etag every 15 minutes if we serialize it?  It seems more kind to have a condition that indicates, "ContactFailed" that is marked after five minutes of being down, before removal.

Reply:
Yes, that's a good point. We should just have a "Stale" entry. I forget why we were proposing a timestamp, maybe we weren't thinking we'd literally be hashing the values which include this for the etag.